File: images/quicksort.cs
using System;
using System.Collections.Generic;
using System.Linq;

class Program
{
    public static List<T> Quicksort<T>(List<T> arr, Comparison<T> comparison)
    {
        if (arr.Count <= 1) {
            return arr;
        }
        T pivot = arr[arr.Count / 2];
        List<T> left = arr.Where(x => comparison(x, pivot) < 0).ToList();
        List<T> middle = arr.Where(x => comparison(x, pivot) == 0).ToList();
        List<T> right = arr.Where(x => comparison(x, pivot) > 0).ToList();
        return Quicksort(left, comparison).Concat(middle).Concat(Quicksort(right, comparison)).ToList();
    }

    static void Main(string[] args)
    {
        List<int> numbers = new List<int>{3,6,8,10,1,2,1};
        List<int> sortedNumbers = Quicksort(numbers, (a, b) => a.CompareTo(b));
        Console.WriteLine(string.Join(" ", sortedNumbers));

        List<string> words = new List<string>{"apple", "grape", "banana", "cherry"};
        List<string> sortedWords = Quicksort(words, (a, b) => a.CompareTo(b));
        Console.WriteLine(string.Join(" ", sortedWords));
    }
}

File: images/quicksort.cpp
#include <iostream>
#include <vector>
#include <string>

template <typename T>
std::vector<T> quicksort(const std::vector<T>& arr) {
    if (arr.size() <= 1) {
        return arr;
    }
    T pivot = arr[arr.size() / 2];
    std::vector<T> left, middle, right;
    for (const T& item : arr) {
        if (item < pivot) left.push_back(item);
        else if (item == pivot) middle.push_back(item);
        else right.push_back(item);
    }
    left = quicksort(left);
    right = quicksort(right);
    left.insert(left.end(), middle.begin(), middle.end());
    left.insert(left.end(), right.begin(), right.end());
    return left;
}

int main() {
    std::vector<int> numbers = {3,6,8,10,1,2,1};
    std::vector<int> sortedNumbers = quicksort(numbers);
    for (const int& num : sortedNumbers) {
        std::cout << num << " ";
    }
    std::cout << std::endl;

    std::vector<std::string> words = {"apple", "grape", "banana", "cherry"};
    std::vector<std::string> sortedWords = quicksort(words);
    for (const std::string& word : sortedWords) {
        std::cout << word << " ";
    }
    std::cout << std::endl;
}
